karapace/kafka_rest_apis/__init__.py

```diff
@@ -739,18 +742,88 @@ def is_valid_schema_request(data: dict, prefix: str) -> bool:
                 return False
         return isinstance(schema, str)
 
-    async def get_schema_id(self, data: dict, topic: str, prefix: str, schema_type: SchemaType) -> int:
-        log.debug("Retrieving schema id for %r", data)
-        if f"{prefix}_schema_id" in data and data[f"{prefix}_schema_id"] is not None:
-            log.debug("Will use schema id %d for serializing %s on topic %s", data[f"{prefix}_schema_id"], prefix, topic)
-            return int(data[f"{prefix}_schema_id"])
-        schema_str = data[f"{prefix}_schema"]
-        log.debug("Registering / Retrieving ID for schema %s", schema_str)
-        if schema_str not in self.schemas_cache:
-            subject_name = self.serializer.get_subject_name(topic, data[f"{prefix}_schema"], prefix, schema_type)
-            schema_id = await self.serializer.get_id_for_schema(data[f"{prefix}_schema"], subject_name, schema_type)
-            self.schemas_cache[schema_str] = schema_id
-        return self.schemas_cache[schema_str]
+    async def get_schema_id(self, data: dict, topic: str, prefix: str, schema_type: SchemaType) -> SchemaId:
+        """
+        This method search and validate the SchemaId for a request, it acts as a guard (In case of something wrong
+        throws an error).
+        """
+        log.debug("[resolve schema id] Retrieving schema id for %r", data)
+        schema_id: Union[SchemaId, None] = (
+            SchemaId(int(data[f"{prefix}_schema_id"])) if f"{prefix}_schema_id" in data else None
+        )
+        schema_str = data.get(f"{prefix}_schema")
+        parsed_schema = None if schema_str is None else ValidatedTypedSchema.parse(schema_type, schema_str)
+        subject_name = self.serializer.get_subject_name(topic, parsed_schema, prefix, schema_type)
+
+        if schema_id is None:
+            schema_id = await self._query_schema_id_from_cache_or_registry(parsed_schema, schema_str, subject_name)
+        else:
+            await self._check_subject_has_schema(schema_id, subject_name, parsed_schema)
+
+        return schema_id
+
+    async def _check_subject_has_schema(
+        self, schema_id: SchemaId, subject_name: Subject, parsed_schema: Union[TypedSchema, None]
```

Comment:
```suggestion
        self, schema_id: SchemaId, subject_name: Subject, parsed_schema: TypedSchema | None
```

Reply:
And also all the other instances

Reply:
I cannot do that, we are targeting python 3.8